sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rule/BeamAggregationRule.java

```diff
@@ -54,29 +61,133 @@ public void onMatch(RelOptRuleCall call) {
 
   private static RelNode updateWindow(RelOptRuleCall call, Aggregate aggregate, Project project) {
     ImmutableBitSet groupByFields = aggregate.getGroupSet();
-    List<RexNode> projectMapping = project.getProjects();
+    ArrayList<RexNode> projects = new ArrayList(project.getProjects());
 
-    Optional<AggregateWindowField> windowField = Optional.empty();
+    WindowFn windowFn = null;
+    int windowFieldIndex = -1;
 
     for (int groupFieldIndex : groupByFields.asList()) {
-      RexNode projNode = projectMapping.get(groupFieldIndex);
+      RexNode projNode = projects.get(groupFieldIndex);
       if (!(projNode instanceof RexCall)) {
         continue;
       }
 
-      windowField = AggregateWindowFactory.getWindowFieldAt((RexCall) projNode, groupFieldIndex);
+      RexCall rexCall = (RexCall) projNode;
+      WindowFn fn = createWindowFn(rexCall.getOperands(), rexCall.op.kind);
+      if (fn != null) {
+        windowFn = fn;
+        windowFieldIndex = groupFieldIndex;
+        projects.set(groupFieldIndex, rexCall.getOperands().get(0));
+      }
     }
 
+    final Project newProject =
+        project.copy(project.getTraitSet(), project.getInput(), projects, project.getRowType());
+
     return new BeamAggregationRel(
         aggregate.getCluster(),
         aggregate.getTraitSet().replace(BeamLogicalConvention.INSTANCE),
-        convert(
-            aggregate.getInput(),
-            aggregate.getInput().getTraitSet().replace(BeamLogicalConvention.INSTANCE)),
+        convert(newProject, newProject.getTraitSet().replace(BeamLogicalConvention.INSTANCE)),
         aggregate.indicator,
         aggregate.getGroupSet(),
         aggregate.getGroupSets(),
         aggregate.getAggCallList(),
-        windowField);
+        windowFn,
+        windowFieldIndex);
+  }
+
+  /**
```

Comment:
Note: here down is moved from `AggregateWindowFactory.java`. This is the only place the code is used and keeping it all together makes it much easier to follow.

Reply:
The two are somewhat intertwined at the moment due to how we apply the `_BEGIN` and `_END` functions. As a result splitting them appears to be non-trivial, so I've opened https://issues.apache.org/jira/browse/BEAM-5416